File: py/aoc2024/day20.py
"""
Day 20: Race Condition
"""

from aoc2024.day20c import solve as _solve

SAMPLE_INPUT = """
###############
#...#...#.....#
#.#.#.#.#.###.#
#S#...#.#.#...#
#######.#.#.###
#######.#.#...#
#######.#.###.#
###..E#...#...#
###.#######.###
#...###...#...#
#.#####.#.###.#
#.#...#.#.#...#
#.#.#.#.#.#.###
#...#...#...###
###############
"""


def _getpath(data: str) -> list[tuple[tuple[int, int], int]]:
    data = data.splitlines()
    pos = next(
        (y, x)
        for y, line in enumerate(data)
        for x, char in enumerate(line)
        if char == "S"
    )
    path = [(pos, 0)]
    last = None
    while True:
        y, x = pos
        if data[y][x] == "E":
            return sorted(path)
        for pos2 in ((y - 1, x), (y, x - 1), (y, x + 1), (y + 1, x)):
            y, x = pos2
            if data[y][x] != "#" and pos2 != last:
                path.append((pos2, len(path)))
                last, pos = pos, pos2
                break
        else:
            return None


def part1(data: str, time: int = 100) -> int:
    """
    >>> counts = [part1(SAMPLE_INPUT, time) for time in (2, 4, 6, 8, 10, 12, 20, 36, 38, 40, 64)]
    >>> [x - y for x, y in zip(counts, counts[1 :] + [0])]
    [14, 14, 2, 4, 2, 3, 1, 1, 1, 1, 1]
    """
    return _solve(_getpath(data), 2, time)


def part2(data: str, time: int = 100) -> int:
    """
    >>> counts = [part2(SAMPLE_INPUT, time) for time in (50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76)]
    >>> [x - y for x, y in zip(counts, counts[1 :] + [0])]
    [32, 31, 29, 39, 25, 23, 20, 19, 12, 14, 12, 22, 4, 3]
    """
    return _solve(_getpath(data), 20, time)


parts = (part1, part2)

File: rs/src/day20.rs
use itertools::Itertools;
use rayon::iter::{IndexedParallelIterator, IntoParallelRefIterator, ParallelIterator};

pub fn part1(data: &str) -> Option<usize> {
    solve(data, 2, 100)
}

pub fn part2(data: &str) -> Option<usize> {
    solve(data, 20, 100)
}

fn solve(data: &str, cheats: usize, time: usize) -> Option<usize> {
    let lines = data.lines().collect::<Vec<_>>();
    let start = lines
        .iter()
        .enumerate()
        .flat_map(|(y, line)| {
            line.bytes()
                .enumerate()
                .filter_map(move |(x, b)| Some((y, x)).filter(|_| b == b'S'))
        })
        .exactly_one()
        .ok()?;
    let mut path = vec![(start, 0)];
    loop {
        let prev = path.iter().nth_back(1).map(|(pos, _)| pos);
        let (pos, b) = path
            .last()
            .iter()
            .flat_map(|((y, x), _)| {
                [
                    (y.wrapping_sub(1), *x),
                    (*y, x.wrapping_sub(1)),
                    (*y, x.wrapping_add(1)),
                    (y.wrapping_add(1), *x),
                ]
            })
            .map(|pos @ (y, x)| (pos, lines.get(y).map(|line| line.as_bytes().get(x))))
            .filter(|(pos, b)| Some(pos) != prev && b != &Some(Some(&b'#')))
            .exactly_one()
            .ok()?;
        path.push((pos, path.len()));
        if b == Some(Some(&b'E')) {
            break;
        }
    }
    path.sort_unstable();
    Some(
        path.par_iter()
            .enumerate()
            .map(|(i, ((y1, x1), t1))| {
                path[i + 1..]
                    .iter()
                    .take_while(|(pos, _)| *pos <= (y1 + cheats, *x1))
                    .filter(|((y2, x2), t2)| {
                        let d = y1.abs_diff(*y2) + x1.abs_diff(*x2);
                        d <= cheats && d + time <= t1.abs_diff(*t2)
                    })
                    .count()
            })
            .sum(),
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use indoc::indoc;
    use pretty_assertions::assert_eq;

    static EXAMPLE: &str = indoc! {"
        ###############
        #...#...#.....#
        #.#.#.#.#.###.#
        #S#...#.#.#...#
        #######.#.#.###
        #######.#.#...#
        #######.#.###.#
        ###..E#...#...#
        ###.#######.###
        #...###...#...#
        #.#####.#.###.#
        #.#...#.#.#...#
        #.#.#.#.#.#.###
        #...#...#...###
        ###############
    "};

    #[test]
    fn part1_examples() {
        [
            (14, 2),
            (14, 4),
            (2, 6),
            (4, 8),
            (2, 10),
            (3, 12),
            (1, 20),
            (1, 36),
            (1, 38),
            (1, 40),
            (1, 64),
        ]
        .into_iter()
        .rev()
        .fold(0, |acc, (count, time)| {
            assert_eq!(Some(acc + count), solve(EXAMPLE, 2, time));
            acc + count
        });
    }

    #[test]
    fn part2_examples() {
        [
            (32, 50),
            (31, 52),
            (29, 54),
            (39, 56),
            (25, 58),
            (23, 60),
            (20, 62),
            (19, 64),
            (12, 66),
            (14, 68),
            (12, 70),
            (22, 72),
            (4, 74),
            (3, 76),
        ]
        .into_iter()
        .rev()
        .fold(0, |acc, (count, time)| {
            assert_eq!(Some(acc + count), solve(EXAMPLE, 20, time));
            acc + count
        });
    }
}
